### PR TITLE
chore: move render code to csaf

### PR DESCRIPTION
**Pull request by bxf12315**

### CODE REVIEW

File: csaf/src/lib.rs [hunk 1]
```diff
@@ -61,5 +61,6 @@ pub mod validation;
 pub mod visitors;
 pub mod walker;
 
+pub mod report;
```
Comment (ctron) [inline csaf/src/lib.rs:64]: Please align that with the previous set of (unconditional) imports

File: csaf/csaf-cli/src/cmd/report/mod.rs [hunk 2]
```diff
@@ -1,10 +1,10 @@
-mod render;
-
 use crate::{
     cmd::{DiscoverArguments, FilterArguments},
     common::walk_visitor,
 };
 use async_trait::async_trait;
+use csaf_walker::report::render;
```
Comment (ctron) [inline csaf/csaf-cli/src/cmd/report/mod.rs:6]: Please used nested `use` sections. Also in the other cases.

File: csaf/Cargo.toml [hunk 3]
```diff
@@ -40,6 +40,8 @@ walker-common = { version = "0.6.0-alpha.7", path = "../common" }
 sequoia-openpgp = { version = "1", default-features = false }
 
 csaf = { version = "0.5", default-features = false, optional = true }
+clap = "4.4.8"
```
Comment (ctron) [inline csaf/Cargo.toml:43]: We should avoid adding `clap` as a dependency to the library version of this crate. So we need to split up the "options" struct into "options" and "arguments" like with the other options as well.

File: csaf/src/report/mod.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+pub mod render;
```
Comment (ctron) [inline csaf/src/report/mod.rs:1]: I think we need to restructure this. Right now there is a module named `crate::report::report`, which doesn't make too much sense to me. It needs to be restructured.

File: csaf/src/report/mod.rs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+use crate::discover::DiscoveredAdvisory;
+use std::borrow::Cow;
+use std::collections::{BTreeMap, HashSet};
+use std::path::PathBuf;
+use url::Url;
+use walker_common::utils::url::Urlify;
+
+#[derive(clap::Args, Debug)]
```
Comment (ctron) [inline csaf/src/report/mod.rs:10]: If we split this up into a clap and plain options variant, please have the `RenderOptions` use the `#[non_exhaustive]` attribute, and use a "builder pattern", like the other options struct now also do.

This allows us to add new fields later, without breaking the API. And I am pretty sure the render options will be changed quite often.

**Review by ctron — CHANGES_REQUESTED** [review]

File: csaf/src/report/mod.rs (hunk 4 above)
Comment (bxf12315) [reply csaf/src/report/mod.rs:1]: it's report::render.  Does it still need to be restructure?

Comment (ctron) [reply csaf/src/report/mod.rs:1]: The fully qualified name will be `crate::report::report::render`, right?

File: csaf/src/report/render.rs [hunk 6]
```diff
@@ -35,7 +38,7 @@ impl Display for Title {
     }
 }
 
-struct HtmlReport<'r>(&'r ReportResult<'r>, &'r RenderOptions);
+struct HtmlReport<'r>(&'r ReportResult<'r>, &'r PathBuf, &'r Option<Url>);
```
Comment (ctron) [inline csaf/src/report/render.rs:41]: As we are adding more fields, I think it makes sense to use a struct variant `{}` instead of tuple.

Comment (bxf12315) [reply csaf/src/report/render.rs:41]: like this csaf-walker::report::render,
https://github.com/trustification/trustification/pull/1064/files#diff-ef9fdbc1979542d4bd996a2209d42c407d7e0255ff2340390c6179a3179ee154R8

File: csaf/src/report/render.rs [hunk 7]
```diff
@@ -0,0 +1,301 @@
+// use crate::report::DocumentKey;
```
Comment (ctron) [inline csaf/src/report/render.rs:1]: Do we still need this file?

File: csaf/src/report/mod.rs [hunk 8]
```diff
@@ -35,7 +77,7 @@ impl Display for Title {
     }
 }
 
-struct HtmlReport<'r>(&'r ReportResult<'r>, &'r RenderOptions);
+struct HtmlReport<'r>(&'r ReportResult<'r>, &'r PathBuf, &'r Option<Url>);
```
Comment (ctron) [inline csaf/src/report/mod.rs:80]: As we're adding too many fields, please switch to the struct form `{}`, using appropriate field names, for example:

```rust
struct HtmlReport<'r> {
  result: …,
  path: …,
  url: …,
}
```

File: csaf/src/lib.rs (hunk 1 above)
Comment (ctron) [reply csaf/src/lib.rs:64]: Please align that with the previous set of (unconditional) imports

File: csaf/csaf-cli/src/cmd/report/mod.rs [hunk 9]
```diff
@@ -1,12 +1,12 @@
-mod render;
-
 use crate::{
     cmd::{DiscoverArguments, FilterArguments},
     common::walk_visitor,
 };
 use async_trait::async_trait;
+use csaf_walker::report::{render_to_html, ReportRenderOption};
```
Comment (ctron) [inline csaf/csaf-cli/src/cmd/report/mod.rs:6]: Please use nested imports.

File: csaf/Cargo.toml (hunk 3 above)
Comment (ctron) [reply csaf/Cargo.toml:43]: This still seems like an open issue: 

> We should avoid adding clap as a dependency to the library version of this crate. So we need to split up the "options" struct into "options" and "arguments" like with the other options as well.

**Review by ctron — CHANGES_REQUESTED** [review]

File: csaf/src/report/mod.rs [hunk 10]
```diff
@@ -35,13 +81,18 @@ impl Display for Title {
     }
 }
 
-struct HtmlReport<'r>(&'r ReportResult<'r>, &'r RenderOptions);
+struct HtmlReport<'r> {
+    result: &'r ReportResult<'r>,
+    #[allow(dead_code)]
```
Comment (ctron) [inline csaf/src/report/mod.rs:86]: If we don't need it, can't we remove it?

File: csaf/Cargo.toml [hunk 11]
```diff
@@ -40,6 +40,7 @@ walker-common = { version = "0.6.0-alpha.7", path = "../common" }
 sequoia-openpgp = { version = "1", default-features = false }
 
 csaf = { version = "0.5", default-features = false, optional = true }
+html-escape = "0.2.13"
```
Comment (ctron) [inline csaf/Cargo.toml:43]: It looks like this is not being used. If so, can it be removed?

**Review by ctron — CHANGES_REQUESTED** [review]
Could you please double check those seemingly unused items. If we can, we should remove them. If we need those, then I think it makes sense adding an explanation.

**Review by ctron — APPROVED** [review]

**Pull request merged**